File: Packages/CaptureKit/Tests/CaptureKitTests/WindowPickingTests.swift
import TestKit
import CoreGraphics
@testable import CaptureKit

private func win(_ id: UInt32, _ frame: CGRect, layer: Int = 0, pid: pid_t = 10,
                 title: String? = nil) -> PickableWindow {
    PickableWindow(id: id, frame: frame, title: title, layer: layer, ownerPID: pid)
}

let windowPickingTests: [TestCase] = [
    TestCase("topmostReturnsFrontOnOverlap") { t in
        // Front-to-back ordered: smaller front window wins over the larger one behind.
        let front = win(1, CGRect(x: 0, y: 0, width: 100, height: 100), pid: 10)
        let back  = win(2, CGRect(x: 0, y: 0, width: 200, height: 200), pid: 11)
        t.equal(WindowPicking.topmost(at: CGPoint(x: 50, y: 50),
                                      windows: [front, back], excludingPID: 99)?.id, 1)
    },
    TestCase("skipsNonNormalLayer") { t in
        let menu = win(1, CGRect(x: 0, y: 0, width: 100, height: 100), layer: 25)   // e.g. menu/dock
        let app  = win(2, CGRect(x: 0, y: 0, width: 100, height: 100), layer: 0, pid: 11, title: "W")
        t.equal(WindowPicking.topmost(at: CGPoint(x: 10, y: 10),
                                      windows: [menu, app], excludingPID: 99)?.id, 2)
    },
    TestCase("excludesOwnPID") { t in
        let own   = win(1, CGRect(x: 0, y: 0, width: 100, height: 100), pid: 42, title: "self")
        let other = win(2, CGRect(x: 0, y: 0, width: 100, height: 100), pid: 7, title: "other")
        t.equal(WindowPicking.topmost(at: CGPoint(x: 10, y: 10),
                                      windows: [own, other], excludingPID: 42)?.id, 2)
    },
    TestCase("missReturnsNil") { t in
        let w = win(1, CGRect(x: 0, y: 0, width: 10, height: 10))
        t.isNil(WindowPicking.topmost(at: CGPoint(x: 500, y: 500),
                                      windows: [w], excludingPID: 99))
    },
    TestCase("cocoaFrameConversion") { t in
        // Primary display 900 tall; window top-left at y=100, height 200 → cocoa y = 900-100-200 = 600.
        let cocoa = WindowPicking.cocoaFrame(
            fromTopLeft: CGRect(x: 50, y: 100, width: 300, height: 200), primaryHeight: 900)
        t.equal(cocoa, CGRect(x: 50, y: 600, width: 300, height: 200))
    },
]

File: windows/tests/BetterScreenshot.Tests/WindowPickingTests.cs
using BetterScreenshot.Capture;
using BetterScreenshot.Core;
using Xunit;

namespace BetterScreenshot.Tests;

public class WindowPickingTests
{
    private static readonly PxPoint Hit = new(50, 50);

    [Fact]
    public void TopmostReturnsFrontOnOverlap()
    {
        var front = new PickableWindow(1, new PxRect(0, 0, 100, 100), "front", 0, 7);
        var back = new PickableWindow(2, new PxRect(0, 0, 500, 500), "back", 0, 7);
        var w = WindowPicking.Topmost(Hit, new[] { front, back }, excludePid: 999);
        Assert.Equal(1u, w!.Value.Id);
    }

    [Fact]
    public void SkipsNonNormalLayer()
    {
        var menu = new PickableWindow(1, new PxRect(0, 0, 100, 100), "menu", 25, 7);
        var app = new PickableWindow(2, new PxRect(0, 0, 100, 100), "app", 0, 7);
        var w = WindowPicking.Topmost(Hit, new[] { menu, app }, excludePid: 999);
        Assert.Equal(2u, w!.Value.Id);
    }

    [Fact]
    public void ExcludesOwnPid()
    {
        var own = new PickableWindow(1, new PxRect(0, 0, 100, 100), "own", 0, 42);
        var other = new PickableWindow(2, new PxRect(0, 0, 100, 100), "other", 0, 7);
        var w = WindowPicking.Topmost(Hit, new[] { own, other }, excludePid: 42);
        Assert.Equal(2u, w!.Value.Id);
    }

    [Fact]
    public void MissReturnsNull()
    {
        var win = new PickableWindow(1, new PxRect(0, 0, 10, 10), "small", 0, 7);
        Assert.Null(WindowPicking.Topmost(new PxPoint(500, 500), new[] { win }, excludePid: 999));
    }
}
